xlt-timerrecorder-chrome/src/background.js

```diff
@@ -234,281 +294,351 @@ function sendTimingDataEntries(tabId, onlyFinished) {
 
-  chrome.tabs.get(details.tabId, function(tab) {
+  chrome.tabs.get(details.tabId, function (tab) {
     if (!chrome.runtime.lastError) {
-      const request = getRequestEntry(details.tabId, details.requestId, details.url);
-      request.responseEnd = details.timeStamp;
-
-      request.error = true;
-      request.finished = true;
-      request.aborted = details.error === "net::ERR_ABORTED";
-      request.method = details.method;
-      request.type = details.type;
-      request.fromCache = details.fromCache;
-      request.responseSize = getResponseSize(details);
-
-      if (request.startTime === null) {
-        request.startTime = details.timeStamp;
-      }
+      getAndModifyRequestEntry(details.tabId, details.requestId, details.url, function (request) {
+        request.responseEnd = details.timeStamp;
+
+        request.error = true;
+        request.finished = true;
+        request.aborted = details.error === "net::ERR_ABORTED";
+        request.method = details.method;
+        request.type = details.type;
+        request.fromCache = details.fromCache;
+        request.responseSize = getResponseSize(details);
+
+        if (request.startTime === null) {
+          request.startTime = details.timeStamp;
+        }
+      });
     }
   });
 }, {
   urls: ["<all_urls>"]
 }, ["extraHeaders"]);
 
-chrome.runtime.onMessage.addListener(function(message, sender, sendResponse) {
+chrome.runtime.onMessage.addListener(async function (message, sender, sendResponse) {
+  await initStorageProm;
+
   const data = getDataIfContentMessage(message);
-  if (!data)
+  if (!data || ["eventBeforeUnload", "eventLoad", "eventResourceTimingBufferFull"].indexOf(data) < 0) {
     return;
+  }
 
   const tabId = sender.tab.id;
+  const TimingData = storageCache["TimingData"] || {};
+
   if (data === "eventBeforeUnload") {
-    if (hasTimingDataEntry(tabId)) {
-      const currentDataEntry = getCurrentTimingDataEntry(tabId);
+    if (hasTimingDataEntry(TimingData, tabId)) {
+      const currentDataEntry = getCurrentTimingDataEntry(TimingData, tabId);
       currentDataEntry.beforeUnload = true;
       setTimingData(currentDataEntry, message.value);
     }
-    newTimingDataEntry(tabId);
+    newTimingDataEntry(TimingData, tabId);
   } else if (data === "eventLoad") {
-    if (hasTimingDataEntry(tabId)) {
-      getCurrentTimingDataEntry(tabId).loaded = true;
+    if (hasTimingDataEntry(TimingData, tabId)) {
+      getCurrentTimingDataEntry(TimingData, tabId).loaded = true;
     }
   }
-  else if(data === "eventResourceTimingBufferFull") {
-    if(hasTimingDataEntry(tabId)) {
-      setTimingData(getCurrentTimingDataEntry(tabId), message.value);
+  else if (data === "eventResourceTimingBufferFull") {
+    if (hasTimingDataEntry(TimingData, tabId)) {
+      setTimingData(getCurrentTimingDataEntry(TimingData, tabId), message.value);
     }
   }
+
+  storeStorageCache();
+});
+
+chrome.runtime.onSuspend.addListener(function () {
+  storeStorageCache();
```

Comment:
Now that use of session storage is optional, what about to log an error here in case service worker is suspended but use of session storage is disabled? Doing so might help to figure out whether period pings prevent the service worker to be suspended or not.

Reply:
Tests have shown that the `onSuspend` event is not fired when the service worker is deactivated. To my knowledge, there is also no other event for that. Looks like they don't want extensions to delay their deactivation in any way.

Reply:
AFAIK this event was introduced to let service workers be notified in case they will be suspended. Thus, it rather seems to be a bug in case this event is never fired.
Anyway, it should not hurt to have an additional log statement here as well; just in case Chrome developers decide to fix the missing event dispatch.